langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonStringWithType.java

```diff
@@ -0,0 +1,71 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package org.ballerinalang.langlib.value;
+
+import org.ballerinalang.jvm.BallerinaErrors;
+import org.ballerinalang.jvm.JSONParser;
+import org.ballerinalang.jvm.StringUtils;
+import org.ballerinalang.jvm.scheduling.Strand;
+import org.ballerinalang.jvm.util.exceptions.BallerinaException;
+import org.ballerinalang.jvm.values.TypedescValue;
+import org.ballerinalang.jvm.values.api.BString;
+import org.ballerinalang.model.types.TypeKind;
+import org.ballerinalang.natives.annotations.Argument;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.ReturnType;
+
+import static org.ballerinalang.jvm.util.exceptions.BallerinaErrorReasons.CONSTRUCT_FROM_CONVERSION_ERROR;
+import static org.ballerinalang.util.BLangCompilerConstants.VALUE_VERSION;
+
+/**
+ * Extern function lang.values:fromJsonWithType.
+ * Converts a string in JSON format to a user-specified type.
+ *
+ * @since 2.0
+ */
+@BallerinaFunction(
+        orgName = "ballerina",
+        packageName = "lang.value", version = VALUE_VERSION,
+        functionName = "fromJsonStringWithType",
+        args = {
+                @Argument(name = "str", type = TypeKind.STRING),
+                @Argument(name = "t", type = TypeKind.TYPEDESC)
+        },
+        returnType = {
+                @ReturnType(type = TypeKind.ANYDATA),
+                @ReturnType(type = TypeKind.ERROR)
+        },
+        isPublic = true
+)
+public class FromJsonStringWithType {
+
+    public static Object fromJsonStringWithType(Strand strand, BString value, TypedescValue t) {
+
+        String str = value.getValue();
+        if (str.equals("null")) {
+            return null;
```

Comment:
We need to check if the target type is nilable.

The folowing crashes.

```ballerina
type IntMap map<int>;

public function main() {
    string s = "null";

    map<int> m = checkpanic s.fromJsonStringWithType(IntMap); // doesn't panic

    int i = m.get("first");
}
```

The get fails with an NPE in the internal log.
```log
[2020-07-16 00:57:59,786] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at org.ballerinalang.langlib.map.Get.get(Get.java:45)
	at ballerina.lang_map.1_1_0.map.get(map.bal:51)
	at qwe.main(qwe.bal:8)
	at ___init.$lambda$main$(.)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:500)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:276)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:244)
	at java.lang.Thread.run(Thread.java:745)
```

Reply:
Do we need to special case "null" anyway? The `JSONParser` and `FromJsonWithType.fromJsonWithType` will handle this properly right?

Reply:
Ack.